core/src/main/java/com/sequenceiq/cloudbreak/service/upgrade/sync/CmSyncImageCollectorService.java

```diff
@@ -0,0 +1,76 @@
+package com.sequenceiq.cloudbreak.service.upgrade.sync;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import javax.inject.Inject;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Service;
+
+import com.sequenceiq.cloudbreak.cloud.model.catalog.Image;
+import com.sequenceiq.cloudbreak.core.CloudbreakImageCatalogException;
+import com.sequenceiq.cloudbreak.core.CloudbreakImageNotFoundException;
+import com.sequenceiq.cloudbreak.domain.stack.Stack;
+import com.sequenceiq.cloudbreak.service.image.ImageCatalogService;
+import com.sequenceiq.cloudbreak.service.image.ImageService;
+import com.sequenceiq.cloudbreak.structuredevent.LegacyRestRequestThreadLocalService;
+
+@Service
+public class CmSyncImageCollectorService {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(CmSyncImageCollectorService.class);
+
+    @Inject
+    private ImageCatalogService imageCatalogService;
+
+    @Inject
+    private ImageService imageService;
+
+    @Inject
+    private LegacyRestRequestThreadLocalService legacyRestRequestThreadLocalService;
+
+    public Set<Image> collectImages(String userCrn, Stack stack, Set<String> candidateImageUuids) {
+        try {
+            String imageCatalogName = imageService.getCurrentImageCatalogName(stack.getId());
+            Long workspaceId = stack.getWorkspace().getId();
+            return candidateImageUuids.isEmpty()
+                    ? getAllImagesFromCatalog(userCrn, stack, imageCatalogName, workspaceId)
+                    : getCurrentAndSelectedImagesFromCatalog(stack, candidateImageUuids, imageCatalogName, workspaceId);
+        } catch (Exception e) {
+            LOGGER.warn("It is not possible to collect images for CM sync, returning empty collection: ", e);
+            return Set.of();
+        }
+    }
+
+    private Set<Image> getCurrentAndSelectedImagesFromCatalog(Stack stack, Set<String> candidateImageUuids, String imageCatalogName, Long workspaceId)
+            throws CloudbreakImageNotFoundException, CloudbreakImageCatalogException {
+        Set<Image> candidateImages = new HashSet<>();
+        candidateImages.addAll(
+                candidateImageUuids.stream()
+                        .map(uuid -> getImage(imageCatalogName, workspaceId, uuid))
+                        .collect(Collectors.toSet())
+        );
+        candidateImages.add(imageService.getCurrentImage(stack.getId()).getImage());
+        return candidateImages;
+    }
+
+    private Set<Image> getAllImagesFromCatalog(String userCrn, Stack stack, String imageCatalogName, Long workspaceId) throws CloudbreakImageCatalogException {
+        legacyRestRequestThreadLocalService.setCloudberUserFromUserCrn(userCrn);
```

Comment:
When asking for all the images at some point the logged in user is retrieved.
https://github.com/hortonworks/cloudbreak/blob/ccae0b6936fffa81553d7c5f6d6084776cc59ad5/core/src/main/java/com/sequenceiq/cloudbreak/service/image/ImageCatalogService.java#L570
I have a jira up to work on this piece, because the current implementation does not give back all the images of a cloud platform, just the last images of each line (i.e. last patch of say 7.2.10). 
https://jira.cloudera.com/browse/CB-13449

If needed, we can add to that jira to use prewarmed images only. I did not do that, however, because maybe in the future CDP will enable the use of base images via a customer image catalog.

Reply:
but why do you need to set the CB user? It should be already set when the call reaches the controller, and it should be available I think through the whole flow, shouldn't it?
Still, if not, I would go by refactoring this a bit, as you would set the user from crn just to get the crn... So create a new method which accepts the crn and call that one

Reply:
The userCrn is passed to the flow, but here CloudbreakUser is needed, as most methods are directly called from api and not from a flow. Yes, a bit indirect just to pass the crn, I refactored it.